### PR TITLE
Fix missing dependencies for Test and optional sections.

### PR DESCRIPTION
**Pull request by gildor478**
1. this change allows to take into account test dependencies that can also
be optional.

For example:
```
Executable test1
  Build$: flag(tests)
  BuildDepends: ounit

Executable test2
  Build$: flag(tests) && flag(long)
  BuildDepends: ounit
```
In this case oasis2opam now generates opam:
```
depends: [
  "ounit" {tests}
]
```
Before this change, "ounit" was a depopts.

2. this change allows to take into account test dependencies from the
TestTools field of Test section.

For example:
```
Test test1
  TestTools: omake, camlp4
```
In this case oasis2opam now generates opam:
```
depends: [
  "camlp4" {tests}
  "omake" {tests}
]
```
Before this change, "omake" and "camlp4" were not taken into account.

### CODE REVIEW

File: src/BuildDepends.ml [hunk 1]
```diff
@@ -65,6 +65,13 @@ let dependency_merge d1 d2 =
     Version.satisfy_both_constraints d1.constraints d2.constraints in
   (* A dep. is compulsory as soon as any occurrence of it is. *)
   let compulsory = d1.compulsory || d2.compulsory in
+  let constraints =
+    let open Version in
+    if not compulsory && (is_test d1.constraints || is_test d2.constraints) then
```
Comment (Chris00) [inline src/BuildDepends.ml]: Is it enough for one of the two to be a test?  It seems to me that if one of the two is not, then it is wrong not to install it if the _test_ flag is not present.

**Review by Chris00 — COMMENTED** [review]
Thanks for the PR.  May you please check the comment?

File: src/BuildDepends.ml (hunk 1 above)
Comment (gildor478) [inline src/BuildDepends.ml]: AFAIK,

compulsory > optional > test

if one of the two is compulsory -> compulsory
the only cases left are:
1. optional, test
2. test, optional
3. test, test 
4. optional, optional

1., 2. and 3. should give "test" and not "optional" and this is covered by the condition.

**Comment by Chris00** [comment]
On 2016-12-05, Sylvain Le Gall wrote:
> 
> AFAIK,
> 
> compulsory > optional > test
> 
> if one of the two is compulsory -> compulsory
> the only cases left are:
> 1. optional, test
> 2. test, optional
> 3. test, test 
> 4. optional, optional
> 
> 1., 2. and 3. should give "test" and not "optional" and this is 
> covered by the condition.

OK, I see your reasoning.  That decision is already taken in 
Version.satisfy_both_constraints however.  I also think that the 
"compulsory" field made sense in the past (before {build} and {test} 
existed) but should now be merged with the above.

**Comment by gildor478** [comment]
Not sure what you want me to do here. Can you give me a more actionable item to work on?

**Comment by Chris00** [comment]
On 2016-12-07, Sylvain Le Gall wrote:
>
> Not sure what you want me to do here. Can you give me a more 
> actionable item to work on?

I pushed a rewritten (better) way of categorizing the dependencies.  It 
would be great if you could test it on you packages and report problems.

**Comment by gildor478** [comment]
Nice, thank you.

**Pull request merged**